Allow using an already unpacked mathjax distribution

Add a MATHJAX_SOURCE_DIR CMake variable, analogous to CLAD_SOURCE_DIR and
OPENUI5_SOURCE_DIR, that points to an already unpacked mathjax
distribution. With it, ROOT can be built with builtin_mathjax=ON without
network access, as the download via ExternalProject and the connectivity
check are skipped.

---
diff --git a/builtins/mathjax/CMakeLists.txt b/builtins/mathjax/CMakeLists.txt
index 773a66cc33b..baa5991a147 100644
--- a/builtins/mathjax/CMakeLists.txt
+++ b/builtins/mathjax/CMakeLists.txt
@@ -1,12 +1,26 @@
-ExternalProject_Add(
-   MATHJAX
-   URL ${lcgpackages}/mathjax-3.2.0.tar.gz
-   URL_HASH SHA256=6847f1cfdd980afbe31b466474da301a56ab859a67d8035442b052dd4ac2fa03
-   CONFIGURE_COMMAND ""
-   BUILD_COMMAND ""
-   INSTALL_COMMAND ""
-   SOURCE_DIR ${CMAKE_BINARY_DIR}/js/mathjax/
-   TIMEOUT 600
-)
+if(DEFINED MATHJAX_SOURCE_DIR)
+  # Use an already unpacked mathjax distribution instead of downloading it,
+  # analogous to CLAD_SOURCE_DIR for clad. This allows to build ROOT without
+  # network access.
+  if(NOT EXISTS "${MATHJAX_SOURCE_DIR}/es5/tex-svg.js")
+    message(FATAL_ERROR "MATHJAX_SOURCE_DIR=${MATHJAX_SOURCE_DIR} does not contain a mathjax distribution (es5/tex-svg.js not found)")
+  endif()
+  file(MAKE_DIRECTORY ${CMAKE_BINARY_DIR}/js)
+  execute_process(COMMAND ${CMAKE_COMMAND} -E create_symlink
+     ${MATHJAX_SOURCE_DIR} ${CMAKE_BINARY_DIR}/js/mathjax)
+  set(_mathjax_dir ${MATHJAX_SOURCE_DIR})
+else()
+  ExternalProject_Add(
+     MATHJAX
+     URL ${lcgpackages}/mathjax-3.2.0.tar.gz
+     URL_HASH SHA256=6847f1cfdd980afbe31b466474da301a56ab859a67d8035442b052dd4ac2fa03
+     CONFIGURE_COMMAND ""
+     BUILD_COMMAND ""
+     INSTALL_COMMAND ""
+     SOURCE_DIR ${CMAKE_BINARY_DIR}/js/mathjax/
+     TIMEOUT 600
+  )
+  set(_mathjax_dir ${CMAKE_BINARY_DIR}/js/mathjax)
+endif()
 
-install(DIRECTORY ${CMAKE_BINARY_DIR}/js/mathjax/ DESTINATION ${CMAKE_INSTALL_JSROOTDIR}/mathjax ${DIR_PERMISSIONS})
+install(DIRECTORY ${_mathjax_dir}/ DESTINATION ${CMAKE_INSTALL_JSROOTDIR}/mathjax ${DIR_PERMISSIONS})
diff --git a/cmake/modules/SearchInstalledSoftware.cmake b/cmake/modules/SearchInstalledSoftware.cmake
index 4af4e2e722e..0de86db5e71 100644
--- a/cmake/modules/SearchInstalledSoftware.cmake
+++ b/cmake/modules/SearchInstalledSoftware.cmake
@@ -1115,10 +1115,12 @@ endif()
 #------------------------------------------------------------------------------------
 if(http)
   if(builtin_mathjax)
-    ROOT_CHECK_CONNECTION("builtin_mathjax=OFF")
-    if (NO_CONNECTION)
-      message(SEND_ERROR "builtin_mathjax=ON requires internet connection, check it or disable feature")
-      list(APPEND HOTFIX_BUILD_FLAGS -Dbuiltin_mathjax=OFF)
+    if(NOT DEFINED MATHJAX_SOURCE_DIR)
+      ROOT_CHECK_CONNECTION("builtin_mathjax=OFF")
+      if (NO_CONNECTION)
+        message(SEND_ERROR "builtin_mathjax=ON requires internet connection, check it or disable feature")
+        list(APPEND HOTFIX_BUILD_FLAGS -Dbuiltin_mathjax=OFF)
+      endif()
     endif()
     add_subdirectory(builtins/mathjax)
   else()
